csrc/preseg_passes/mark_aliases_prepare.cpp

```diff
@@ -66,19 +92,60 @@ std::unordered_set<Expr*> exprsDependedByNonAliases(
   return {depended_by_non_aliases.begin(), depended_by_non_aliases.end()};
 }
 
-// Inserts a `segment_set` after `use_of` and redirect aliasing users to
-// use the `segment_set`.
-void insertSegmentSetAfter(
-    std::vector<Use>::const_iterator first_user,
-    std::vector<Use>::const_iterator last_user) {
+// Inserts a `segment_set` after `use_of` to separate meta and non-meta ops.
+template <typename InputIter>
+void insertSegmentSetAfter(InputIter first_user, InputIter last_user) {
   TensorView* use_of = first_user->use_of;
 
-  // There are a few corner cases where we don't need to add a
-  // `segment_set`. If `use_of` is only used by aliases, ...
-  if (static_cast<size_t>(std::distance(first_user, last_user)) ==
-      use_of->uses().size()) {
+  std::vector<Expr*> users;
+  users.reserve(use_of->uses().size());
+  // `uses_to_segment` is sorted so `nullptr` if exists appears first.
+  if (first_user->user == nullptr) {
+    // This is an optimization to make fewer segments. In the
+    // following example, if bookending wants to segment (a) between `use_of`
+    // and all its users, (b) between `use_of` and `user_0`, and (c) between
+    // `use_of` and `user_1`. We can instead segment only between `use_of` and
+    // `user_2`, the complement set of [`first_user`, `last_user`). This is
+    // valid because the ops before `use_of`, those after `user_0`, and those
+    // after `user_1` are all meta ops that can be merged into one no-op
+    // segment.
+    //
+    //   use_of | -> | [user 0]
+    //            |
+    //            +> | [user 1]
+    //            |
+    //            +> [user 2]
+    //
+    //   ==>
+    //
+    //   use_of -> [user 0]
+    //          |
+    //          +> [user 1]
+    //          |
+    //          +> | [user 2]
+    first_user++;
+    std::unordered_set<Expr*> to_remove;
+    std::for_each(first_user, last_user, [&](const Use& use) {
+      to_remove.insert(use.user);
+    });
+    std::copy_if(
+        use_of->uses().begin(),
+        use_of->uses().end(),
+        std::back_inserter(users),
+        [&](Expr* user) { return to_remove.count(user) == 0; });
+  } else {
+    std::transform(
+        first_user, last_user, std::back_inserter(users), [](const Use& use) {
+          return use.user;
+        });
+  }
+
+  // There are a few corner cases where we can avoid adding a
+  // `segment_set`. If a segment_set is to be added between `use_of` and all
+  // its users, ...
+  if (users.size() == use_of->uses().size()) {
```

Comment:
The reason we check for this is to use `segment_seg` to avoid horizontal fusion of common consumers of a fusion input?

Reply:
This is for cases like: 

```
in --> [meta 0] -> out0
   |
   +-> [meta 1] -> out1
```

No `segment_set`s are needed at all.